zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplineServeTask.kt

```diff
@@ -50,18 +56,44 @@ abstract class ZiplineServeTask : DefaultTask() {
     val deploymentRegistry = services.get(DeploymentRegistry::class.java)
     val deploymentHandle = deploymentRegistry.get(deploymentId, ZiplineServerDeploymentHandle::class.java)
     if (deploymentHandle == null) {
-      val server = routes(
-        "/" bind static(Directory(inputDir.get().asFile.absolutePath), Pair("zipline", ContentType.TEXT_PLAIN))
-      ).asServer(SunHttp(port.orElse(8080).get()))
-
+      // First time this task is run, start the server
+      val ws = websockets(
+        "/ws" bind { ws: Websocket ->
+          websockets.add(ws)
+          ws.send(WsMessage(RELOAD_MESSAGE))
+          ws.onClose {
+            websockets.remove(ws)
+          }
+        }
+      )
+      val http = routes(
+        "/" bind static(
+          Directory(inputDir.get().asFile.absolutePath),
+          Pair("zipline", ContentType.TEXT_PLAIN)
+        )
+      )
+      val server = PolyHandler(http, ws).asServer(Jetty(8080))
       deploymentRegistry.start(
         deploymentId,
         DeploymentRegistry.ChangeBehavior.BLOCK,
         ZiplineServerDeploymentHandle::class.java,
         server
       )
+
+      // Keep the connection open by sending a message periodically
+      Timer("WebsocketHeartbeat", true).schedule(0, 10000) {
+        websockets.forEach { it.send(WsMessage("heartbeat")) }
```

Comment:
This should probably be a ping instead. OkHttp will actually automatically ping from the other side to determine if the connection is alive, but I suppose we don't want to rely on that.

Reply:
Looks like OkHttp doesn't do this automatically but there is a [setting](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/-builder/ping-interval/) to turn it on. The connection was closing every 30 seconds previously which is why I added the heartbeat. 

I'm not sure if proper ping-pong really matters here. In the current implementation the client websocket will be closed if we don't receive a message every 30 seconds, and the server side should receive a close message most of the time when the websocket closes. There may be a slow resource leak over time where the server keeps some connections open that it shouldn't but given that this is a hot reload server we'll be restarting it constantly anyway and clearing out the old connections. That being said, I'm not particularly opposed to to ping-pong instead of heartbeat, but it's just a bit more code complication.

Thoughts?

Reply:
Is that our implementation that gives up after 30 seconds? Or is that just a default timeout?

A ping is nice because it never reaches the user code. Both clients and servers must respond automatically to pings with a pong to conform with the spec. And if it's an automatic OkHttp timeout that automatically closes the connection, the ping will keep it alive automatically.

Reply:
Ah, I didn't realize that it was part of the spec, that makes more sense. Apparently http4k does not conform though. It won't respond to OkHttp's pings. Looks like it doesn't [properly map frames ](https://github.com/http4k/http4k/blob/master/http4k-server/jetty/src/main/kotlin/org/http4k/server/Http4kWebSocketFrameHandler.kt#L29)to the backing Jetty. Womp womp.

I think the 30 seconds may be a server idle timeout, but I don't see an easy way to configure. 

So we're left with the heartbeat as the simplest way to make things work, even if it's not quite the proper way of doing things. Definitely less impressed with http4k at this point, though still seems fine for our particular use case. I'll just proceed with this approach unless you have objections.

Reply:
Yeah it also doesn't allow sending pings to which OkHttp would automatically reply (the spec is basically symmetric).

> Definitely less impressed with http4k at this point

Yep, same.

> though still seems fine for our particular use case. I'll just proceed with this approach unless you have objections.

Yep. Merge away.